File: golang/tokenize/tokenize_v2_test.go
package tokenize

import (
	"reflect"
	"testing"

	"github.com/muthuishere/citenexus/golang/internal/conform"
)

// The ADR-0011 per-script golden fixture. The CLAIM (SupportedScripts) and the
// EVIDENCE for it (conformance/cases/tokenize_v2.json) are checked against each
// other here: no script may be claimed as supported without a golden fixture.
type tokenizeV2Fixture struct {
	TokenizerVersion  int      `json:"tokenizer_version"`
	SupportedScripts  []string `json:"supported_scripts"`
	ContinuousScripts []string `json:"continuous_scripts"`
	UnrelatedPassage  string   `json:"unrelated_passage"`
	Supported         []struct {
		Script             string   `json:"script"`
		Text               string   `json:"text"`
		Tokens             []string `json:"tokens"`
		V1Tokens           []string `json:"v1_tokens"`
		SelfSupported      bool     `json:"self_supported"`
		UnrelatedSupported bool     `json:"unrelated_supported"`
		UnsupportedScripts []string `json:"unsupported_scripts"`
	} `json:"supported"`
	Unclaimed []struct {
		Script             string   `json:"script"`
		Text               string   `json:"text"`
		UnsupportedScripts []string `json:"unsupported_scripts"`
	} `json:"unclaimed"`
	Unicode []struct {
		Input  string   `json:"input"`
		Tokens []string `json:"tokens"`
	} `json:"unicode"`
}

func loadTokenizeV2(t *testing.T) tokenizeV2Fixture {
	t.Helper()
	var f tokenizeV2Fixture
	conform.Case(t, "tokenize_v2.json", &f)
	return f
}

// expectedTokenizeV2Counts pins every bucket in
// conformance/cases/tokenize_v2.json. Iterating a bucket asserts nothing about a
// vector that was silently dropped from it.
var expectedTokenizeV2Counts = map[string]int{
	"supported":          14,
	"unclaimed":          11,
	"unicode":            27,
	"supported_scripts":  14,
	"continuous_scripts": 7,
}

func TestTokenizeV2VectorBucketSizes(t *testing.T) {
	f := loadTokenizeV2(t)
	for name, got := range map[string]int{
		"supported":          len(f.Supported),
		"unclaimed":          len(f.Unclaimed),
		"unicode":            len(f.Unicode),
		"supported_scripts":  len(f.SupportedScripts),
		"continuous_scripts": len(f.ContinuousScripts),
	} {
		if want := expectedTokenizeV2Counts[name]; got != want {
			t.Errorf("tokenize_v2.json bucket %q: got %d vectors, want %d", name, got, want)
		}
	}
}

func TestTokenizeV2FixturePinsTheVersion(t *testing.T) {
	if got := loadTokenizeV2(t).TokenizerVersion; got != TokenizerVersion {
		t.Fatalf("tokenizer_version = %d, want %d", got, TokenizerVersion)
	}
}

// The embedded scripts.json is a COPY of the canonical claim in
// conformance/cases/tokenize_v2.json (ADR-0010 tier 2: one definition, generated
// copies, a test that pins them together). This is the Go mirror of
// TestPolarityTableMatchesConformance in golang/gate.
func TestScriptTableMatchesConformance(t *testing.T) {
	f := loadTokenizeV2(t)
	if got := SupportedScripts(); !reflect.DeepEqual(got, f.SupportedScripts) {
		t.Fatalf("embedded scripts.json diverged from conformance:\n got  %v\n want %v", got, f.SupportedScripts)
	}
	if got := ContinuousScripts(); !reflect.DeepEqual(got, f.ContinuousScripts) {
		t.Fatalf("embedded scripts.json diverged from conformance:\n got  %v\n want %v", got, f.ContinuousScripts)
	}
	claimed := map[string]bool{}
	for _, s := range f.SupportedScripts {
		claimed[s] = true
	}
	for _, c := range f.Supported {
		if !claimed[c.Script] {
			t.Fatalf("fixture carries a case for unclaimed script %q", c.Script)
		}
		delete(claimed, c.Script)
	}
	if len(claimed) != 0 {
		t.Fatalf("claimed scripts with no golden fixture: %v", claimed)
	}
}

// A mis-sorted table would silently mis-classify via the binary search.
func TestScriptRangesAreSorted(t *testing.T) {
	for i := 1; i < len(scriptRanges); i++ {
		if scriptRanges[i-1].last >= scriptRanges[i].first {
			t.Fatalf("script ranges must be sorted and disjoint: %v then %v", scriptRanges[i-1], scriptRanges[i])
		}
	}
}

func TestEveryClaimedScriptTokenizes(t *testing.T) {
	for _, c := range loadTokenizeV2(t).Supported {
		got := TokenizeV2(c.Text)
		if !reflect.DeepEqual(got, c.Tokens) {
			t.Fatalf("%s: TokenizeV2 = %q, want %q", c.Script, got, c.Tokens)
		}
		if len(c.Tokens) == 0 {
			t.Fatalf("%s: fixture claims support but expects zero tokens", c.Script)
		}
	}
}

// v1 must NOT be "fixed" — the shipped vectors and every index already built
// under it depend on the ASCII behavior.
func TestV1DefectStaysPinned(t *testing.T) {
	for _, c := range loadTokenizeV2(t).Supported {
		got := Tokenize(c.Text)
		want := c.V1Tokens
		if want == nil {
			want = []string{}
		}
		if !reflect.DeepEqual(got, want) {
			t.Fatalf("%s: v1 Tokenize = %q, want %q", c.Script, got, want)
		}
	}
}

func TestClaimedScriptsReportNoCapabilityGap(t *testing.T) {
	for _, c := range loadTokenizeV2(t).Supported {
		if got := UnsupportedScripts(c.Text); len(got) != 0 {
			t.Fatalf("%s: UnsupportedScripts = %v, want []", c.Script, got)
		}
		if len(c.UnsupportedScripts) != 0 {
			t.Fatalf("%s: fixture disagrees with itself", c.Script)
		}
	}
}

func TestUnclaimedScriptsAreReportedAsACapabilityGap(t *testing.T) {
	f := loadTokenizeV2(t)
	if len(f.Unclaimed) == 0 {
		t.Fatal("the unclaimed half of the matrix must not be empty")
	}
	for _, c := range f.Unclaimed {
		got := UnsupportedScripts(c.Text)
		if !reflect.DeepEqual(got, c.UnsupportedScripts) {
			t.Fatalf("%s: UnsupportedScripts = %v, want %v", c.Script, got, c.UnsupportedScripts)
		}
		if !reflect.DeepEqual(c.UnsupportedScripts, []string{c.Script}) {
			t.Fatalf("%s: fixture disagrees with itself: %v", c.Script, c.UnsupportedScripts)
		}
	}
}

func TestUnicodeMechanicsVectors(t *testing.T) {
	for _, c := range loadTokenizeV2(t).Unicode {
		got := TokenizeV2(c.Input)
		if !reflect.DeepEqual(got, c.Tokens) {
			t.Fatalf("input %q: TokenizeV2 = %q, want %q", c.Input, got, c.Tokens)
		}
	}
}

// v2 is a strict superset of v1 on pure-ASCII input — that equivalence is why
// moving BM25 and the gate onto v2 left every shipped vector unchanged.
func TestV2AgreesWithV1OnAsciiVectors(t *testing.T) {
	var cases []struct {
		Input  string   `json:"input"`
		Tokens []string `json:"tokens"`
	}
	conform.Case(t, "tokenize.json", &cases)
	if len(cases) != expectedTokenizeCases {
		t.Fatalf("tokenize.json: got %d cases, want %d", len(cases), expectedTokenizeCases)
	}
	for _, c := range cases {
		if !isASCIIWords(c.Input) {
			continue
		}
		if got := TokenizeV2(c.Input); !reflect.DeepEqual(got, Tokenize(c.Input)) {
			t.Fatalf("input %q: v2 = %q diverges from v1 = %q on ASCII input", c.Input, got, Tokenize(c.Input))
		}
	}
}

func isASCIIWords(s string) bool {
	for _, r := range s {
		if r > 127 {
			return false
		}
	}
	return true
}

// Telugu (U+0C00-U+0C7F) was absent from the range table entirely: it read as a
// NEIGHBOUR plus "unknown" and still emitted six delimited tokens, so BM25
// ranked a script no fixture had ever validated while the answer flow filtered
// every Telugu passage out of the grounding set.
func TestTeluguClassifiesAsItselfAndIsDelimited(t *testing.T) {
	const text = "ఉద్యోగి రహస్య సమాచారాన్ని"
	if got := ScriptsIn(text); !reflect.DeepEqual(got, []string{"telugu"}) {
		t.Fatalf("ScriptsIn = %q, want [telugu]", got)
	}
	if got := UnsupportedScripts(text); len(got) != 0 {
		t.Fatalf("UnsupportedScripts = %q, want none", got)
	}
	want := []string{"ఉద్యోగి", "రహస్య", "సమాచారాన్ని"}
	if got := TokenizeV2(text); !reflect.DeepEqual(got, want) {
		t.Fatalf("TokenizeV2 = %q, want %q (Telugu writes spaces; bigrams would be the mirror defect)", got, want)
	}
}

// Every CLAIMED script must classify to ITSELF — a second script in this list is
// exactly what Telugu's ('devanagari', 'unknown') was.
func TestEveryClaimedScriptSampleClassifiesToItself(t *testing.T) {
	for _, c := range loadTokenizeV2(t).Supported {
		for _, s := range ScriptsIn(c.Text) {
			// Japanese genuinely mixes Han and Hiragana; nothing may be "unknown".
			if s == "unknown" {
				t.Fatalf("%s: sample classifies partly as unknown", c.Script)
			}
			if s != c.Script && !(c.Script == "hiragana" && s == "han") {
				t.Fatalf("%s: sample also classifies as %q", c.Script, s)
			}
		}
	}
}

// The structural half: a script ABSENT from the range table has no validated
// segmentation rule, so it produces NOTHING. BM25 cannot rank it, and the gate
// cannot accept it (an empty claim never aligns).
func TestAScriptAbsentFromTheTableDoesNotTokenize(t *testing.T) {
	for _, text := range []string{"የሰራተኛው ሚስጥራዊ መረጃ", "ᏗᏙᎳᏅᏍᏗ ᎠᏓᏅᏙ", "ཞིབ་འཇུག"} {
		if got := UnsupportedScripts(text); !reflect.DeepEqual(got, []string{"unknown"}) {
			t.Fatalf("%q: UnsupportedScripts = %q, want [unknown]", text, got)
		}
		if got := TokenizeV2(text); len(got) != 0 {
			t.Fatalf("%q: TokenizeV2 = %q, want no tokens", text, got)
		}
	}
	// Only the unknown RUN is dropped — a stray character cannot silence an
	// otherwise-supported passage.
	want := []string{"2026", "policy"}
	if got := TokenizeV2("የሰራተኛው 2026 policy"); !reflect.DeepEqual(got, want) {
		t.Fatalf("TokenizeV2 = %q, want %q", got, want)
	}
}

File: js/src/tokenize/tokenize-v2.test.ts
// The ADR-0011 per-script golden matrix. The CLAIM (SUPPORTED_SCRIPTS) and the
// EVIDENCE for it (conformance/cases/tokenize_v2.json) are checked against each
// other here: no script may be claimed as supported without a golden fixture.
import { describe, expect, it } from "vitest";

import { loadCase } from "../conform/fixtures.js";
import { isSupportedV2 } from "../gate/verify-v2.js";
import { tokenize } from "./tokenize.js";
import {
  CONTINUOUS_SCRIPTS,
  SCRIPT_RANGES_FOR_TEST,
  SUPPORTED_SCRIPTS,
  TOKENIZER_VERSION,
  tokenizeV2,
  unsupportedScripts,
} from "./tokenize-v2.js";

interface ScriptCase {
  script: string;
  text: string;
  tokens: string[];
  v1_tokens: string[];
  self_supported: boolean;
  unrelated_supported: boolean;
  unsupported_scripts: string[];
}

interface TokenizeV2Fixture {
  tokenizer_version: number;
  supported_scripts: string[];
  continuous_scripts: string[];
  unrelated_passage: string;
  supported: ScriptCase[];
  unclaimed: { script: string; text: string; unsupported_scripts: string[] }[];
  unicode: { input: string; tokens: string[] }[];
}

const FIXTURE = loadCase<TokenizeV2Fixture>("tokenize_v2.json");

describe("tokenizeV2 conformance (ADR-0011)", () => {
  it("the fixture pins the tokenizer version", () => {
    expect(FIXTURE.tokenizer_version).toBe(TOKENIZER_VERSION);
  });

  it("the claim and its evidence are the same artifact", () => {
    expect([...SUPPORTED_SCRIPTS].sort()).toEqual(FIXTURE.supported_scripts);
    expect([...CONTINUOUS_SCRIPTS].sort()).toEqual(FIXTURE.continuous_scripts);
    expect(new Set(FIXTURE.supported.map((c) => c.script))).toEqual(new Set(SUPPORTED_SCRIPTS));
  });

  // A mis-sorted table would silently mis-classify via the binary search.
  it("the script range table is sorted and disjoint", () => {
    for (let i = 1; i < SCRIPT_RANGES_FOR_TEST.length; i += 1) {
      expect(SCRIPT_RANGES_FOR_TEST[i - 1]![1]).toBeLessThan(SCRIPT_RANGES_FOR_TEST[i]![0]);
    }
  });

  it.each(FIXTURE.supported)("$script tokenizes", (c) => {
    expect(tokenizeV2(c.text)).toEqual(c.tokens);
    expect(c.tokens.length).toBeGreaterThan(0);
  });

  // v1 must NOT be "fixed" — the shipped vectors and every index already built
  // under it depend on the ASCII behavior.
  it.each(FIXTURE.supported)("$script keeps the v1 defect pinned", (c) => {
    expect(tokenize(c.text)).toEqual(c.v1_tokens);
  });

  it.each(FIXTURE.supported)("$script reports no capability gap", (c) => {
    expect(unsupportedScripts(c.text)).toEqual([]);
    expect(c.unsupported_scripts).toEqual([]);
  });

  // Both halves must hold or the script is not supported: the gate accepts a
  // verbatim quote of its own source, AND it still rejects unrelated text.
  it.each(FIXTURE.supported)("$script supports a verbatim quote of its own source", (c) => {
    expect(c.self_supported).toBe(true);
    expect(isSupportedV2(c.text, c.text)).toBe(true);
  });

  it.each(FIXTURE.supported)("$script does not turn the gate into a rubber stamp", (c) => {
    expect(c.unrelated_supported).toBe(false);
    expect(isSupportedV2(c.text, FIXTURE.unrelated_passage)).toBe(false);
  });

  it("the unclaimed half of the matrix is not empty", () => {
    expect(FIXTURE.unclaimed.length).toBeGreaterThan(0);
  });

  it.each(FIXTURE.unclaimed)("$script is reported as a capability gap", (c) => {
    expect(SUPPORTED_SCRIPTS.has(c.script)).toBe(false);
    expect(unsupportedScripts(c.text)).toEqual(c.unsupported_scripts);
    expect(c.unsupported_scripts).toEqual([c.script]);
  });

  it.each(FIXTURE.unicode)("unicode mechanics: $input", (c) => {
    expect(tokenizeV2(c.input)).toEqual(c.tokens);
  });

  // v2 is a strict superset of v1 on pure-ASCII input — that equivalence is why
  // moving BM25 and the gate onto v2 left every shipped vector unchanged.
  it("agrees with v1 on every ASCII v1 vector", () => {
    const cases = loadCase<{ input: string; tokens: string[] }[]>("tokenize.json");
    expect(cases.length).toBeGreaterThan(0);
    for (const c of cases) {
      if (Array.from(c.input).some((ch) => ch.codePointAt(0)! > 127)) continue;
      expect(tokenizeV2(c.input)).toEqual(tokenize(c.input));
    }
  });
});
